aspnet/mvc/controllers/filters.rst

```diff
@@ -1,8 +1,312 @@
+
+For an ``IAsyncActionFilter`` the ``OnActionExecutionAsync`` combines all the possibilites of ``OnActionExecuting`` and ``OnActionExecuted``. A call to ``await next()`` on the ``ActionExecutionDelegate`` will execute any subsequent action filters and the action method, returning an ``ActionExecutedContext``. To short-circuit inside of an ``OnActionExecutionAsync``, set ``ActionExecutingContext.Result`` and do not call the ``ActionExectionDelegate``.
+
+The ``DurationActionFilter`` shown below uses a ``Stopwatch`` to measure the execution time of the action, for display on the page. The ``ViewData`` property of the ``ViewResult`` can be accessed and manipulated, including adding new values.
+
+.. _duration-action-filter:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/DurationActionFilter.cs
+  :language: c#
+  :emphasize-lines: 1,21-26
+  :lines: 8-35
+  :dedent: 4
+
+.. note:: If you just need to log how long an action takes to execute, rather than displaying it on the page, this is already done by the framework.
+
+.. _exception-filters:
+
+Exception Filters
+-----------------
+*Exception Filters* implement either the ``IExceptionFilter`` or ``IAsyncExceptionFilter`` interface.
+
+Exception filters handle unhandled exceptions. They are only called when an exception occurs later in the pipeline. They can provide a single location to implement common error handling policies within an app. The framework provides an abstract `ExceptionFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/ExceptionFilterAttribute/index.html>`_ that you should be able to subclass for your needs. Exception filters are good for trapping exceptions that occur within MVC actions, but they're not as flexible as error handling middleware. Prefer middleware for the general case, and use filters only where the middleware doesn't offer the functionality your app requires.
+
+Exception filters do not have two events (for before and after) - they only implement ``OnException`` (or ``OnExceptionAsync``). The ``ExceptionContext`` provided in the ``OnException`` parameter includes the ``Exception`` that occurred. If you set ``context.Exception`` to null, the effect is that you've handled the exception, so the request will proceed as if it hadn't occurred (most likely returning a 200 OK status). The following filter logs exceptions and marks them as handled:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/LoggingExceptionFilterAttribute.cs
+  :language: c#
+  :emphasize-lines: 7,15,18
+
+.. _result-filters:
+
+Result Filters
+--------------
+*Result Filters* implement either the ``IResultFilter`` or ``IAsyncResultFilter`` interface and their execution surrounds the execution of action results. Result filters are only executed for successful results - when the action or action filters produce an action result. **Result filters are not executed when exception filters handle an exception.**
```

Comment:
Can you tie this to Razor views? The OnResultExecuting method runs before the razor view and OnResultExecuted after the razor view completes? Or is it more complicated than that - if so that might be worth mentioning.

I think of action filters as applying to action methods and results filters applying to what's going on in the razor view.